### PR TITLE
hacked in lens and hlint suggestions

### PR DESCRIPTION
**Pull request by metaml**
1. due to S3Event being so deeply nested, i hacked in Lens
2. other hacks due to hlint suggestions

i was a bit aggressive in hacking the code; please feel free to ignore this pr or let me know what changes you want to make this pr acceptable to you.

### CODE REVIEW

File: src/AWS/Lambda/KinesisDataStreamsEvent.hs [hunk 1]
```diff
@@ -1,69 +1,56 @@
...
-    <*> v .: "data"
-    <*> v .: "kinesisSchemaVersion"
-    <*> v .: "sequenceNumber"
+  parseJSON = genericParseJSON defaultOptions { fieldLabelModifier = drop 1 }
 
-instance ToJSON Kinesis where
-  toJSON Kinesis{..} =
-    object [
-        "partitionKey" .= partitionKey
-      , "data" .= payload
```
Comment (amarrella) [inline src/AWS/Lambda/KinesisDataStreamsEvent.hs]: Notice that what we called `payload` in `Kinesis` is serialized/deserialized as "data". We couldn't use the `data` keyword as it's a reserved keyword in the language

File: stack.yaml [hunk 2]
```diff
@@ -4,7 +4,7 @@ packages:
 - .
 
 docker:
-  enable: true
+  enable: false
```
Comment (amarrella) [inline stack.yaml]: why this change? we want to build inside docker to have the same linux installation as AWS Lambda

File: src/AWS/Lambda/KinesisDataStreamsEvent.hs [hunk 3]
```diff
@@ -1,69 +1,56 @@
...
 
 instance ToJSON KinesisDataStreamsEvent where
-  toJSON (KinesisDataStreamsEvent{..}) =
-    object ["Records" .= records]
+  toJSON = genericToJSON defaultOptions { fieldLabelModifier = drop 1 }
+
+instance FromJSON KinesisDataStreamsEvent where
+  parseJSON = genericParseJSON defaultOptions { fieldLabelModifier = drop 1 }
+
+newtype KinesisDataStreamsEvent = KinesisDataStreamsEvent { _records :: [Record] }
```
Comment (amarrella) [inline src/AWS/Lambda/KinesisDataStreamsEvent.hs]: this needs to be uppercase in the JSON

File: src/AWS/Lambda/S3Event.hs [hunk 4]
```diff
@@ -1,94 +1,88 @@
...
-    s3Object        :: S3Object
-  } deriving (Show)
 
-instance FromJSON S3 where
-  parseJSON (Object v) = S3
-    <$> v .: "s3SchemaVersion"
-    <*> v .: "configurationId"
-    <*> v .: "bucket"
-    <*> ((v .: "object") >>= parseJSON)
+newtype S3Event = S3Event { _records :: [Record] }
```
Comment (amarrella) [inline src/AWS/Lambda/S3Event.hs]: `records` needs to be uppercase in the JSON

File: stack.yaml (hunk 2 above)
Comment (metaml) [reply stack.yaml]: "make build" has "--docker" so that builds in docker but i also want "make buildc" for local development. i can go either way on this--so i can revert if you don't find it necessary.

File: src/AWS/Lambda/RuntimeClient.hs [hunk 5]
```diff
@@ -97,9 +97,8 @@ endpoints host =
   where baseURL' = "http://" <> host <> "/2018-06-01/runtime"
 
 forceMaybe :: String -> Maybe a -> a
-forceMaybe errorMsg maybe' = case maybe' of
-  Just a' -> a'
-  Nothing -> error errorMsg
+forceMaybe errorMsg = fromMaybe (error errorMsg)
+--forceMaybe errorMsg maybe' = fromMaybe (error errorMsg) maybe'
```
Comment (louoso) [inline src/AWS/Lambda/RuntimeClient.hs]: This can be removed, right?

File: Makefile [hunk 6]
```diff
@@ -1,14 +1,21 @@
 BUILD_DIR=build
 EXAMPLES=api-gateway kinesis s3
 
-.PHONY: default
```
Comment (louoso) [inline Makefile]: Why remove the `.PHONY` targets?

File: src/AWS/Lambda/KinesisDataStreamsEvent.hs (hunk 3 above)
Comment (metaml) [reply src/AWS/Lambda/KinesisDataStreamsEvent.hs]: Sorry, I'm not sure what "this" is referring to.

File: src/AWS/Lambda/RuntimeClient.hs (hunk 5 above)
Comment (metaml) [reply src/AWS/Lambda/RuntimeClient.hs]: yeah, it was for reference only--to show the eta reduction.

File: Makefile (hunk 6 above)
Comment (metaml) [reply Makefile]: I found them noisy and they're only necessary if the target name is the same as file or directory name. 

I can put them back, if prefer that.

**Comment by louoso** [comment]
Thanks for the PR @metaml. I have a couple of asks, and a couple of questions.

Would you mind if I refactored this into multiple commits and then opened different PRs for each?

Seems like we have:

1) adding lenses
2) adding hlint
3) some makefile additions
4) not building in a docker container

A few questions:

1) Why the change to `stack.yaml`?
2) Is there a nice way to use `fieldModifier` that works for `Records` and `data` fields?
3) For the lenses, I've noticed that most libraries I've used have written the lenses manually rather than use `makeLenses`. Is the reason just to make using lenses optional while keeping the non-lens path user friendly? Is that a better approach?

File: src/AWS/Lambda/KinesisDataStreamsEvent.hs (hunk 3 above)
Comment (amarrella) [reply src/AWS/Lambda/KinesisDataStreamsEvent.hs]: `records`

File: stack.yaml (hunk 2 above)
Comment (louoso) [reply stack.yaml]: Let's do a quick huddle on this on slack. I want to figure out what would be "least astonishing" so having feedback from both of you would be helpful.

**Comment by metaml** [comment]
I don't mind at all. It's your repo, so I'll take whatever you think is acceptable to you. Thanks!

1. Why the stack.yaml?
I want to build locally in mac os, I have a pretty dumb setup where I use "make buildc" same as "stack build --filewatch." For that, I just to build without docker. But when we build for reals, "make build" uses docker.

2. Is there a nice way to use fieldModifier that works for Records and data fields?
They're pretty flexible and should work for both.

3. For the lenses, I've noticed that most libraries... 
It's your call but given the nesting of S3Event, it'll be nice to offer it. The only difference is that if you don't want to use lens, you still have the field accessors except that they're prefixed with a "_".
